integrations/cdi/eclipselink-cdi/src/main/java/io/helidon/integrations/cdi/eclipselink/CDISEPlatform.java

```diff
@@ -183,15 +189,64 @@ public void launchContainerRunnable(final Runnable runnable) {
      * ServerPlatformBase#initializeExternalTransactionController()
      * superclass implementation}.
      *
+     * <p>This method also acquires a {@link DataSource} from
+     * {@linkplain CDI CDI} proactively, and {@linkplain
+     * JNDIConnector#setDataSource(DataSource) installs it} to
+     * pre&euml;mpt any JNDI operations.</p>
+     *
+     * @exception ValidationException if a {@link DataSource} could
+     * not be acquired
+     *
      * @see ServerPlatformBase#initializeExternalTransactionController()
+     *
+     * @see DatabaseSession#getLogin()
+     *
+     * @see DatabaseLogin#getConnector()
+     *
+     * @see JNDIConnector
+     *
+     * @see JNDIConnector#getName()
+     *
+     * @see JNDIConnector#setDataSource(DataSource)
      */
     @Override
     public void initializeExternalTransactionController() {
         final CDI<Object> cdi = CDI.current();
-        if (cdi == null || cdi.select(TransactionManager.class).isUnsatisfied()) {
+        assert cdi != null;
+        if (cdi.select(TransactionManager.class).isUnsatisfied()) {
             this.disableJTA();
         }
         super.initializeExternalTransactionController();
+
+        // See https://github.com/oracle/helidon/issues/949.  This is
+        // the only spot where we can actually change the Connector
+        // that is used by EclipseLink to look up a data source during
+        // JPA "SE mode" persistence unit acquisition such that it
+        // doesn't get overwritten by other EclipseLink internals.
+        final DatabaseSession session = this.getDatabaseSession();
+        if (session != null) {
+            final DatabaseLogin login = session.getLogin();
+            if (login != null) {
+                final Connector connector = login.getConnector();
+                if (connector instanceof JNDIConnector) {
+                    final JNDIConnector jndiConnector = (JNDIConnector) connector;
+                    final String dataSourceName = jndiConnector.getName();
+                    if (dataSourceName == null) {
+                        throw ValidationException.invalidDataSourceName(null, null);
+                    }
+                    final DataSource dataSource;
+                    DataSource temp = null;
+                    try {
+                        temp = cdi.select(DataSource.class, NamedLiteral.of(jndiConnector.getName())).get();
```

Comment:
Can't you just move the call to `jndiConnector.setDataSource(dataSource)` here (and remove the temp and finally block)?

Reply:
Yeah, good catch